azerothcore/azerothcore-wotlk
# Fix issue 5342 mob speed health

## Changes Proposed:
-  call Unit::UpdateSpeed() more often for NPCs
-  in UpdateSpeed(), reduce speed based on health

## Issues Addressed:

- Fixes part of  https://github.com/azerothcore/azerothcore-wotlk/issues/5342

## SOURCE:

See 5342

## Tests Performed:

- builds ok. Tested on my own machine (win10 x64)
- Get in game, find a mob, aggro, reduce hp and check his speed.
- video of my testing : https://drive.google.com/file/d/18Gq1EjtGGVFzkdCg9HNVSNZFDVLnTIuu/view?usp=sharing 
https://drive.google.com/file/d/1v8iruv8V3jqWyLwBR2HdT2ZysesTwRbO/view?usp=sharing 

## How to Test the Changes:

Use macro to read speed. The following one seems to work
/script DEFAULT_CHAT_FRAME:AddMessage(string.format("%f", GetUnitSpeed("target")), 1.0, 1.0, 1.0);
.modify speed 2.5 on yourself will help get away from mobs.
.damage X will be useful to get mobs to accurate hp.
1. Find any mob. Get in combat, have him run at normal speed towards you. Read speed, should be 8 for average mob
2. Damage him to 31%, make sure he's running with no obstacles. Read speed, should be same as above.
3. Damage to anything below 30%, mob starts slowing down. At 1% he should be at about 45% slow (speed of ~4.5 if base is 8)
4. Heal mob, you can use .aura 33763 to cast lifebloom rank 1. speed should increase back to 100% as life reaches > 30%.
5. Do same as above but cast a frostbolt or other slowing spell, slow should be more than before

Maybe review we didn't affect players or pets in any wrong way, but the code should be warded against that.

Second part of the bug:
1. Find a mob that runs away at 15% hp.
2. Get him to ~17%, read speed.
3. Drop him just below 15 so he runs away, read speed. Should be comparable (within 5%).
4. Regardless of the mob finding someone or not to assist, his speed shouldn't change.

## Known Issues and TODO List:

No known issue, but I made a change down in movement manager (of something that should probably not have been there in the first place), could maybe cause odd things.


## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.
